### PR TITLE
SDLP and R1CS BP shared inputs

### PR DESCRIPTION
**Pull request by ryanorendorff**
Enables pieces of the BP R1CS implementation to be shared with the SDLP portion. Note that the PR for the `sunscreen_bulletproofs` (https://github.com/Sunscreen-tech/bulletproofs_zkcrypto/pull/1) must be merged first.

An example program linking the BFV SDLP and a balance check can be seen here. The next PR will be making this into a pair of functions (prove and verify) that users can call more easily (with inputs being the LatticeProblem, the ZKP, and the ZKP inputs).

```rust
use bitvec::vec::BitVec;
use curve25519_dalek::{ristretto::RistrettoPoint, scalar::Scalar};
use merlin::Transcript;
use sunscreen::bulletproofs::{BulletproofProverParameters, BulletproofVerifierParameters};
use sunscreen::types::zkp::BigInt;
use sunscreen::{
    bulletproofs::BulletproofsBackend,
    types::zkp::{BulletproofsField, ConstrainCmp, Field, FieldSpec, ProgramNode},
    zkp_program, zkp_var, Compiler, Error, SecurityLevel, ZkpProgramInput, ZkpRuntime,
};
use sunscreen_bulletproofs::{BulletproofGens, PedersenGens};

use seal_fhe::{
    BFVScalarEncoder, BfvEncryptionParametersBuilder, CoefficientModulus, Context, Encryptor,
    KeyGenerator, PlainModulus, PolynomialArray,
};

use logproof::{
    linear_algebra::Matrix,
    math::{make_poly, next_higher_power_of_two, rand256},
    rings::{SealQ128_1024, ZqRistretto},
    test::{
        bfv_delta, convert_to_polynomial, convert_to_smallint, strip_trailing_value, LatticeProblem,
    },
    Bounds, LogProof, LogProofGenerators, LogProofProverKnowledge, LogProofTranscript,
};

use sunscreen_math::{
    poly::Polynomial,
    ring::{BarrettBackend, BarrettConfig, Zq},
    Zero,
};

fn from_twos_complement_field_element<F: FieldSpec, const N: usize>(
    x: [ProgramNode<Field<F>>; N],
) -> ProgramNode<Field<F>> {
    let mut x_recon = zkp_var!(0);

    for (i, x_i) in x.iter().enumerate().take(N - 1) {
        x_recon = x_recon + (zkp_var!(2i64.pow(i as u32)) * (*x_i));
    }

    x_recon = x_recon + zkp_var!(-(2i64.pow((N - 1) as u32))) * x[N - 1];

    x_recon
}

fn from_twos_complement_int(x: &BitVec) -> i64 {
    let mut x_recon = 0;

    for i in 0..(x.len() - 1) {
        x_recon += 2i64.pow(i as u32) * x[i] as i64;
    }

    x_recon -= 2i64.pow((x.len() - 1) as u32) * x[x.len() - 1] as i64;

    x_recon
}

// Can represent programs up to a certain value
#[zkp_program]
fn valid_transaction<F: FieldSpec>(
    #[private] x: [Field<F>; 15],
    #[private] a: Field<F>,
    #[public] balance: Field<F>,
) {
    let lower_bound = zkp_var!(0);

    // Reconstruct x from the bag of bits
    let x_recon = from_twos_complement_field_element(x);

    // Constraint that x is less than or equal to balance
    balance.constrain_ge_bounded(x_recon, 64);

    // Constraint that x is greater than or equal to zero
    lower_bound.constrain_le_bounded(x_recon, 64);

    balance.constrain_ge_bounded(a, 64);
}

fn test_seal_linear_relation<B, const N: usize>(
    message: u64,
    degree: u64,
    plain_modulus: u64,
) -> LatticeProblem<Zq<N, BarrettBackend<N, B>>>
where
    B: BarrettConfig<N>,
{
    let degree = degree;

    let plain_modulus = PlainModulus::raw(plain_modulus).unwrap();
    let coeff_modulus = CoefficientModulus::bfv_default(degree, SecurityLevel::TC128).unwrap();

    // Calculate the data coefficient modulus, which for fields with more
    // that one modulus in the coefficient modulus set is equal to the
    // product of all but the last moduli in the set.
    let mut data_modulus = ZqRistretto::from(1);

    if coeff_modulus.len() == 1 {
        data_modulus = data_modulus * ZqRistretto::from(coeff_modulus[0].value());
    } else {
        for modulus in coeff_modulus.iter().take(coeff_modulus.len() - 1) {
            data_modulus = data_modulus * ZqRistretto::from(modulus.value());
        }
    }

    // Generate encryption parameters and encrypt/decrypt functions.
    let params = BfvEncryptionParametersBuilder::new()
        .set_poly_modulus_degree(degree)
        .set_coefficient_modulus(coeff_modulus.clone())
        .set_plain_modulus(plain_modulus.clone())
        .build()
        .unwrap();

    let ctx = Context::new(&params, false, SecurityLevel::TC128).unwrap();
    let gen = KeyGenerator::new(&ctx).unwrap();

    let encoder = BFVScalarEncoder::new();

    let public_key = gen.create_public_key();
    let secret_key = gen.secret_key();

    let encryptor = Encryptor::with_public_and_secret_key(&ctx, &public_key, &secret_key).unwrap();

    // Generate plaintext data
    let plaintext = encoder.encode_unsigned(message).unwrap();

    // Generate an encrypted message with components
    let (ciphertext, u_exported, e_exported, r_exported) = encryptor
        // .encrypt_return_components(&plaintext, true, None)
        .encrypt_return_components(&plaintext)
        .unwrap();

    // Convert all components into their polynomial representations in the
    // fields we use in this package.
    let m = Polynomial {
        coeffs: strip_trailing_value(
            (0..plaintext.len())
                .map(|i| Zq::from(plaintext.get_coefficient(i)))
                .collect::<Vec<_>>(),
            Zq::zero(),
        ),
    };

    let u = convert_to_polynomial(u_exported.clone()).pop().unwrap();
    let u_small = convert_to_smallint(&coeff_modulus, u_exported.clone());

    let mut es = convert_to_polynomial(e_exported.clone());
    let e_1 = es.remove(0);
    let e_2 = es.remove(0);
    let e_small = convert_to_smallint(&coeff_modulus, e_exported.clone());

    let mut cs =
        convert_to_polynomial(PolynomialArray::new_from_ciphertext(&ctx, &ciphertext).unwrap());
    let c_0 = cs.remove(0);
    let c_1 = cs.remove(0);

    let mut pk =
        convert_to_polynomial(PolynomialArray::new_from_public_key(&ctx, &public_key).unwrap());
    let p_0 = pk.remove(0);
    let p_1 = pk.remove(0);

    let r_coeffs = (0..r_exported.len())
        .map(|i| r_exported.get_coefficient(i))
        .collect::<Vec<u64>>();
    let r = Polynomial {
        coeffs: r_coeffs
            .iter()
            .map(|r_i| Zq::from(*r_i))
            .collect::<Vec<_>>(),
    };

    // Delta is the constant polynomial with floor (q/t) as it's DC compopnent.
    let delta_dc = bfv_delta(data_modulus, plain_modulus.value());
    let delta_dc = Zq::try_from(delta_dc).unwrap();

    let delta = Polynomial {
        coeffs: vec![delta_dc],
    };

    // Set up the BFV equations.
    let one = make_poly(&[1]);
    let zero = make_poly(&[]);

    let a = Matrix::<Polynomial<_>>::from([
        [
            delta.clone(),
            one.clone(),
            p_0.clone(),
            one.clone(),
            zero.clone(),
        ],
        [
            zero.clone(),
            zero.clone(),
            p_1.clone(),
            zero.clone(),
            one.clone(),
        ],
    ]);

    let s = Matrix::<Polynomial<_>>::from([
        [m.clone()],
        [r.clone()],
        [u.clone()],
        [e_1.clone()],
        [e_2.clone()],
    ]);

    // Set up the field polymonial divisor (x^N + 1).
    let mut f_components = vec![0; (degree + 1) as usize];
    f_components[0] = 1;
    f_components[degree as usize] = 1;
    let f = make_poly(&f_components);

    // We do this without the polynomial division and then perform that at
    // the end.
    let mut t = &a * &s;

    // Divide back to a polynomial of at max degree `degree`
    let t_0 = t[(0, 0)].vartime_div_rem_restricted_rhs(&f).1;
    let t_1 = t[(1, 0)].vartime_div_rem_restricted_rhs(&f).1;
    t[(0, 0)] = t_0;
    t[(1, 0)] = t_1;

    // Test that our equations match the matrix result.
    let t_0_from_eq = (delta * &m).vartime_div_rem_restricted_rhs(&f).1
        + r.clone()
        + (p_0 * &u).vartime_div_rem_restricted_rhs(&f).1
        + e_1;

    let t_1_from_eq = (p_1 * &u).vartime_div_rem_restricted_rhs(&f).1 + e_2;

    // Assertions that the SEAL ciphertext matches our calculated one. We
    // use panics here to avoid the large printout from assert_eq.

    if t[(0, 0)] != t_0_from_eq {
        panic!("Matrix and written out equation match for t_0");
    }

    if t[(1, 0)] != t_1_from_eq {
        panic!("Matrix and written out equation match for t_1");
    }

    if t[(0, 0)] != c_0 {
        panic!("t_0 and c_0 are not equal");
    }

    if t[(1, 0)] != c_1 {
        panic!("t_1 and c_1 are not equal");
    }

    // Assert that the equations are equal when written up as a matrix (this
    // should trivially pass if the above assertions pass)
    assert_eq!(t, Matrix::<Polynomial<_>>::from([[c_0], [c_1]]));

    // Calculate bounds for the zero knowledge proof
    let mut m_coeffs = (0..plaintext.len())
        .map(|i| plaintext.get_coefficient(i) as i64)
        .collect::<Vec<i64>>();

    m_coeffs.resize(degree as usize, 0);

    let mut r_coeffs = (0..r_exported.len())
        .map(|i| r_exported.get_coefficient(i) as i64)
        .collect::<Vec<i64>>();

    r_coeffs.resize(degree as usize, 0);

    let s_components = vec![
        m_coeffs,
        r_coeffs,
        u_small[0].clone(),
        e_small[0].clone(),
        e_small[1].clone(),
    ];

    let s_components_bounds = s_components
        .into_iter()
        .map(|v| {
            Bounds(
                v.into_iter()
                    .map(|x| {
                        if x == 0 {
                            0
                        } else {
                            next_higher_power_of_two(x.unsigned_abs())
                        }
                    })
                    .collect::<Vec<u64>>(),
            )
        })
        .collect::<Vec<Bounds>>();

    let b: Matrix<Bounds> = Matrix::from(s_components_bounds);

    LatticeProblem { a, s, t, f, b }
}

fn zero_knowledge_proof<B, const N: usize>(
    message: u64,
    degree: u64,
    plain_modulus: u64,
    transcript: &mut Transcript,
) -> (LogProof, BitVec, Vec<RistrettoPoint>, Scalar)
where
    B: BarrettConfig<N>,
{
    let LatticeProblem { a, s, t, f, b } =
        test_seal_linear_relation::<B, N>(message, degree, plain_modulus);

    let pk = LogProofProverKnowledge::new(&a, &s, &t, &b, &f);

    let total_bits = pk.vk.b()[(0, 0)].iter().sum::<u64>() as usize;

    let s_binary = pk.s_binary_by_index((0, 0));

    // We will not verify for now, verify on joint
    let mut verify_transcript = transcript.clone();

    let gens = LogProofGenerators::new(pk.vk.l() as usize);
    let u = PedersenGens::default().B_blinding;

    // Note that this is half rho
    let half_rho = Scalar::from_bits(rand256());
    let proof =
        LogProof::create_with_shared(transcript, &pk, &gens.g, &gens.h, &u, &half_rho, &[(0, 0)]);

    proof
        .verify(&mut verify_transcript, &pk.vk, &gens.g, &gens.h, &u)
        .unwrap();

    let l = transcript.challenge_scalar(b"verify");
    let r = verify_transcript.challenge_scalar(b"verify");

    assert_eq!(l, r);

    (proof, s_binary, gens.h[0..total_bits].to_vec(), half_rho)
}

fn main() -> Result<(), Error> {
    let app = Compiler::new()
        .zkp_backend::<BulletproofsBackend>()
        .zkp_program(valid_transaction)
        .compile()?;

    let valid_transaction_zkp = app.get_zkp_program(valid_transaction).unwrap();

    let runtime = ZkpRuntime::new(BulletproofsBackend::new())?;

    // in binary:  010111011011111
    //             010111011011111
    let x = 11999u64;
    let mut transcript = Transcript::new(b"valid_encryption_and_zkp");

    // Prove the encryption is valid
    let (proof_sdlp, msg_bits, shared_gens, half_rho) =
        zero_knowledge_proof::<SealQ128_1024, 1>(x, 1024, 12289, &mut transcript);
    println!("Proved encryption");

    println!("msg bits: {:?}", &msg_bits);
    println!("x_bp as int: {}", from_twos_complement_int(&msg_bits));

    let x_bp: [BulletproofsField; 15] = msg_bits
        .iter()
        .map(|x| BulletproofsField::from(*x as u64))
        .collect::<Vec<_>>()
        .try_into()
        .unwrap();
    let balance = BulletproofsField::from(12000u64);

    let transcript_sdlp = transcript.clone();

    let private_inputs_start: Vec<ZkpProgramInput> =
        vec![x_bp.into(), BulletproofsField::from(15u64).into()];
    let public_inputs_start: Vec<ZkpProgramInput> = vec![balance.into()];
    let constant_inputs_start: Vec<ZkpProgramInput> = vec![];

    let private_inputs = private_inputs_start
        .iter()
        .flat_map(|x| x.0.to_native_fields())
        .collect::<Vec<BigInt>>();
    let public_inputs = public_inputs_start
        .iter()
        .flat_map(|x| x.0.to_native_fields())
        .collect::<Vec<BigInt>>();
    let constant_inputs = constant_inputs_start
        .iter()
        .flat_map(|x| x.0.to_native_fields())
        .collect::<Vec<BigInt>>();

    let backend = BulletproofsBackend::new();
    let metrics = backend
        .metrics(
            valid_transaction_zkp,
            &private_inputs,
            &public_inputs,
            &constant_inputs,
        )
        .unwrap();

    let constraint_count = backend
        .constraint_count(
            valid_transaction_zkp,
            &private_inputs,
            &public_inputs,
            &constant_inputs,
        )
        .unwrap();

    let bulletproof_gens = BulletproofGens::new_single_party_with_shared_generators(
        2 * constraint_count,
        &shared_gens,
        metrics.multipliers - 1,
        metrics.pending_multiplier,
    );

    let verifier_parameters = BulletproofVerifierParameters::new(
        PedersenGens::default(),
        bulletproof_gens.clone(),
        shared_gens.len(),
    );

    let prover_parameters = BulletproofProverParameters::new(verifier_parameters.clone(), half_rho);

    let proof = runtime.prove_with_parameters::<ZkpProgramInput>(
        valid_transaction_zkp,
        private_inputs_start,
        public_inputs_start,
        constant_inputs_start,
        &prover_parameters,
        &mut transcript_sdlp.clone(),
    )?;

    println!("Proved ZKP with shared input");

    if let sunscreen::Proof::Bulletproofs(ref b) = proof {
        let b = b.clone();
        let a_i1_shared = (*b).0.A_I1_shared();

        if a_i1_shared != proof_sdlp.w_shared.compress() {
            println!("Shared commitments are not the same!!!!!!!!!!!!!!!!!!!!!!!1");
        } else {
            println!("Shared commitments in SDLP and R1CS BP are the same");
        }
    }

    // Verify the proof
    runtime.verify_with_parameters(
        valid_transaction_zkp,
        &proof,
        vec![balance],
        vec![],
        &verifier_parameters,
        &mut transcript_sdlp.clone(),
    )?;

    println!("Verified ZKP with shared input");

    println!("All proofs pass and are verified!");

    Ok(())
}
```

### CODE REVIEW

File: logproof/src/linear_relation.rs [hunk 1]
```diff
@@ -189,6 +195,37 @@ where
             .sum()
     }
 
+    /**
+     * Rannges in the serialized coefficients of S corresponding to the bounds
```
Comment (rickwebiii) [inline logproof/src/linear_relation.rs:199]: Ranges

File: logproof/src/linear_relation.rs [hunk 2]
```diff
@@ -333,6 +370,23 @@ where
 
         Self { s: s.clone(), vk }
     }
+
+    /// Currently we only really use this to pull out the first n bits, but we
```
Comment (rickwebiii) [inline logproof/src/linear_relation.rs:374]: This isn't a very good doc comment :)

Comment (ryanorendorff) [reply logproof/src/linear_relation.rs:374]: Fixed to mention that the index is into S.

File: logproof/src/linear_relation.rs (hunk 1 above)
Comment (ryanorendorff) [reply logproof/src/linear_relation.rs:199]: Thanks!

File: logproof/src/linear_relation.rs [hunk 3]
```diff
@@ -859,6 +975,62 @@ impl LogProof {
...
+        u: &RistrettoPoint,
+    ) -> RistrettoPoint {
+        let mut commitment = RistrettoPoint::identity();
+
+        let s_1_shared = s_1_shared.iter().map(|b| b == true).collect::<Vec<bool>>();
+
+        commitment += s_1_shared
+            .par_iter()
+            .zip_eq(h.par_iter())
+            .filter_map(|(bit, p_i)| if *bit { Some(p_i) } else { None })
```
Comment (rickwebiii) [inline logproof/src/linear_relation.rs:991]: Is this worth parallelizing? I don't know what the scheduling overhead is.

Comment (ryanorendorff) [reply logproof/src/linear_relation.rs:996]: It's what was there before (when the commitment was generated at once) but it is not precisely a long operation. So I don't think parallelization is buying us a lot.

Comment (ryanorendorff) [reply logproof/src/linear_relation.rs:996]: Are you making a comment because the shared part is likely to be much smaller than the non-shared part?

**Review by rickwebiii — APPROVED** [review]

**Pull request merged**